File: src/bus/broker.py
"""Session bus broker — SQLite-backed cross-project messaging.

Persistent broker for CruxDev sessions to discover each other,
report issues, share patterns, and notify breaking changes.
All state in a single SQLite database at ~/.cruxdev/bus.db.
"""

from __future__ import annotations

import json
import os
import sqlite3
import time
import uuid
from contextlib import contextmanager
from dataclasses import dataclass
from enum import Enum
from typing import Optional


DEFAULT_DB_PATH = os.path.join(
    os.environ.get("HOME", ""),
    ".cruxdev",
    "bus.db",
)


class MessageType(Enum):
    ISSUE = "issue"
    IMPROVEMENT = "improvement"
    PATTERN = "pattern"
    BREAKING_CHANGE = "breaking_change"
    CUSTOM = "custom"


class MessageSeverity(Enum):
    HIGH = "high"
    MEDIUM = "medium"
    LOW = "low"


@dataclass
class Session:
    id: str
    project: str
    directory: str
    registered_at: float
    last_heartbeat: float


@dataclass
class Message:
    id: str
    type: str
    source_project: str
    target_project: str  # "*" for broadcast
    title: str
    body: str
    severity: str
    created_at: float
    acknowledged: bool = False
    acknowledged_at: Optional[float] = None


class Broker:
    """SQLite-backed message broker for cross-project session communication."""

    def __init__(self, db_path: str = DEFAULT_DB_PATH):
        self.db_path = db_path
        os.makedirs(os.path.dirname(db_path), exist_ok=True)
        self._init_db()

    @contextmanager
    def _conn(self):
        conn = sqlite3.connect(self.db_path)
        conn.row_factory = sqlite3.Row
        try:
            yield conn
            conn.commit()
        finally:
            conn.close()

    def _init_db(self) -> None:
        with self._conn() as conn:
            conn.execute("""
                CREATE TABLE IF NOT EXISTS sessions (
                    id TEXT PRIMARY KEY,
                    project TEXT NOT NULL,
                    directory TEXT NOT NULL,
                    registered_at REAL NOT NULL,
                    last_heartbeat REAL NOT NULL
                )
            """)
            conn.execute("""
                CREATE TABLE IF NOT EXISTS messages (
                    id TEXT PRIMARY KEY,
                    type TEXT NOT NULL,
                    source_project TEXT NOT NULL,
                    target_project TEXT NOT NULL,
                    title TEXT NOT NULL,
                    body TEXT NOT NULL,
                    severity TEXT NOT NULL DEFAULT 'medium',
                    created_at REAL NOT NULL,
                    acknowledged INTEGER NOT NULL DEFAULT 0,
                    acknowledged_at REAL
                )
            """)
            conn.execute("""
                CREATE INDEX IF NOT EXISTS idx_messages_target
                ON messages(target_project, acknowledged)
            """)

    # --- Session management ---

    def register_session(self, project: str, directory: str) -> str:
        """Register a session. Returns session ID."""
        session_id = str(uuid.uuid4())[:8]
        now = time.time()
        with self._conn() as conn:
            conn.execute(
                "INSERT INTO sessions (id, project, directory, registered_at, last_heartbeat) VALUES (?, ?, ?, ?, ?)",
                (session_id, project, directory, now, now),
            )
        return session_id

    def heartbeat(self, session_id: str) -> None:
        """Update session heartbeat."""
        with self._conn() as conn:
            conn.execute(
                "UPDATE sessions SET last_heartbeat = ? WHERE id = ?",
                (time.time(), session_id),
            )

    def unregister_session(self, session_id: str) -> None:
        """Remove a session."""
        with self._conn() as conn:
            conn.execute("DELETE FROM sessions WHERE id = ?", (session_id,))

    def list_sessions(self, max_age_seconds: float = 3600) -> list[Session]:
        """List active sessions (heartbeat within max_age)."""
        cutoff = time.time() - max_age_seconds
        with self._conn() as conn:
            rows = conn.execute(
                "SELECT * FROM sessions WHERE last_heartbeat > ? ORDER BY project",
                (cutoff,),
            ).fetchall()
        return [
            Session(
                id=r["id"],
                project=r["project"],
                directory=r["directory"],
                registered_at=r["registered_at"],
                last_heartbeat=r["last_heartbeat"],
            )
            for r in rows
        ]

    def cleanup_stale_sessions(self, max_age_seconds: float = 3600) -> int:
        """Remove sessions with no heartbeat within max_age. Returns count removed."""
        cutoff = time.time() - max_age_seconds
        with self._conn() as conn:
            cursor = conn.execute(
                "DELETE FROM sessions WHERE last_heartbeat < ?",
                (cutoff,),
            )
        return cursor.rowcount

    # --- Messaging ---

    def send_message(
        self,
        type: str,
        source_project: str,
        target_project: str,
        title: str,
        body: str,
        severity: str = "medium",
    ) -> str:
        """Send a message. Returns message ID.

        Also writes a notification file so hook scripts can detect
        new messages without querying SQLite.
        """
        msg_id = str(uuid.uuid4())[:8]
        with self._conn() as conn:
            conn.execute(
                "INSERT INTO messages (id, type, source_project, target_project, title, body, severity, created_at) VALUES (?, ?, ?, ?, ?, ?, ?, ?)",
                (msg_id, type, source_project, target_project, title, body, severity, time.time()),
            )
        # Write notification file for push-based detection
        if target_project != "*":
            self._write_notification(target_project, title)
        else:
            # Broadcast — notify all registered projects
            for session in self.list_sessions():
                if session.project != source_project:
                    self._write_notification(session.project, title)
        return msg_id

    def _write_notification(self, project: str, title: str) -> None:
        """Write a notification file for a project."""
        notify_dir = os.path.join(os.path.dirname(self.db_path), "notifications")
        os.makedirs(notify_dir, exist_ok=True)
        notify_path = os.path.join(notify_dir, f"{project}.notify")

        # Read existing count if any
        count = 0
        try:
            with open(notify_path) as f:
                existing = json.loads(f.read())
                count = existing.get("count", 0)
        except (FileNotFoundError, json.JSONDecodeError, ValueError):
            pass

        data = json.dumps({
            "count": count + 1,
            "latest_title": title,
            "updated_at": time.time(),
        })
        # Atomic write
        tmp_path = notify_path + ".tmp"
        with open(tmp_path, "w") as f:
            f.write(data)
        os.replace(tmp_path, notify_path)

    def clear_notification(self, project: str) -> bool:
        """Clear the notification file for a project. Returns True if file existed."""
        notify_dir = os.path.join(os.path.dirname(self.db_path), "notifications")
        notify_path = os.path.join(notify_dir, f"{project}.notify")
        try:
            os.remove(notify_path)
            return True
        except FileNotFoundError:
            return False

    def read_notification(self, project: str) -> dict | None:
        """Read notification file for a project. Returns None if no notification."""
        notify_dir = os.path.join(os.path.dirname(self.db_path), "notifications")
        notify_path = os.path.join(notify_dir, f"{project}.notify")
        try:
            with open(notify_path) as f:
                return json.loads(f.read())
        except (FileNotFoundError, json.JSONDecodeError):
            return None

    def check_inbox(
        self,
        project: str,
        include_acknowledged: bool = False,
    ) -> list[Message]:
        """Get messages for a project (including broadcasts)."""
        with self._conn() as conn:
            if include_acknowledged:
                rows = conn.execute(
                    "SELECT * FROM messages WHERE target_project IN (?, '*') ORDER BY created_at DESC",
                    (project,),
                ).fetchall()
            else:
                rows = conn.execute(
                    "SELECT * FROM messages WHERE target_project IN (?, '*') AND acknowledged = 0 ORDER BY created_at DESC",
                    (project,),
                ).fetchall()
        return [self._row_to_message(r) for r in rows]

    def acknowledge(self, message_id: str) -> bool:
        """Mark a message as acknowledged. Returns True if found."""
        with self._conn() as conn:
            cursor = conn.execute(
                "UPDATE messages SET acknowledged = 1, acknowledged_at = ? WHERE id = ?",
                (time.time(), message_id),
            )
        return cursor.rowcount > 0

    def get_message(self, message_id: str) -> Optional[Message]:
        """Get a single message by ID."""
        with self._conn() as conn:
            row = conn.execute(
                "SELECT * FROM messages WHERE id = ?",
                (message_id,),
            ).fetchone()
        return self._row_to_message(row) if row else None

    def get_all_messages(
        self,
        type_filter: Optional[str] = None,
        limit: int = 100,
    ) -> list[Message]:
        """Get all messages, optionally filtered by type."""
        with self._conn() as conn:
            if type_filter:
                rows = conn.execute(
                    "SELECT * FROM messages WHERE type = ? ORDER BY created_at DESC LIMIT ?",
                    (type_filter, limit),
                ).fetchall()
            else:
                rows = conn.execute(
                    "SELECT * FROM messages ORDER BY created_at DESC LIMIT ?",
                    (limit,),
                ).fetchall()
        return [self._row_to_message(r) for r in rows]

    def _row_to_message(self, row: sqlite3.Row) -> Message:
        return Message(
            id=row["id"],
            type=row["type"],
            source_project=row["source_project"],
            target_project=row["target_project"],
            title=row["title"],
            body=row["body"],
            severity=row["severity"],
            created_at=row["created_at"],
            acknowledged=bool(row["acknowledged"]),
            acknowledged_at=row["acknowledged_at"],
        )

    # --- Convenience methods ---

    def report_issue(
        self,
        source_project: str,
        target_project: str,
        title: str,
        body: str,
        severity: str = "medium",
    ) -> str:
        """Report an issue found in another project."""
        return self.send_message(
            type=MessageType.ISSUE.value,
            source_project=source_project,
            target_project=target_project,
            title=title,
            body=body,
            severity=severity,
        )

    def report_improvement(
        self,
        source_project: str,
        target_project: str,
        title: str,
        body: str,
    ) -> str:
        """Suggest an improvement to another project."""
        return self.send_message(
            type=MessageType.IMPROVEMENT.value,
            source_project=source_project,
            target_project=target_project,
            title=title,
            body=body,
            severity="medium",
        )

    def share_pattern(
        self,
        source_project: str,
        pattern_name: str,
        description: str,
    ) -> str:
        """Broadcast a learned pattern to all sessions."""
        return self.send_message(
            type=MessageType.PATTERN.value,
            source_project=source_project,
            target_project="*",
            title=pattern_name,
            body=description,
            severity="low",
        )

    def notify_breaking_change(
        self,
        source_project: str,
        affected_projects: list[str],
        description: str,
    ) -> list[str]:
        """Notify affected projects of a breaking change."""
        msg_ids = []
        for project in affected_projects:
            msg_id = self.send_message(
                type=MessageType.BREAKING_CHANGE.value,
                source_project=source_project,
                target_project=project,
                title=f"Breaking change in {source_project}",
                body=description,
                severity="high",
            )
            msg_ids.append(msg_id)
        return msg_ids

File: rust/src/bus/broker.rs
//! SQLite-backed message broker for cross-project session communication.

use rusqlite::{Connection, params};
use serde::{Deserialize, Serialize};
use std::path::{Path, PathBuf};
use std::fs;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Session {
    pub id: String,
    pub project: String,
    pub directory: String,
    pub registered_at: f64,
    pub last_heartbeat: f64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Message {
    pub id: String,
    #[serde(rename = "type")]
    pub msg_type: String,
    pub source_project: String,
    pub target_project: String,
    pub title: String,
    pub body: String,
    pub severity: String,
    pub created_at: f64,
    pub acknowledged: bool,
}

pub struct Broker {
    db_path: PathBuf,
}

impl Broker {
    pub fn new(db_path: &str) -> anyhow::Result<Self> {
        let path = PathBuf::from(db_path);
        if let Some(parent) = path.parent() {
            fs::create_dir_all(parent)?;
        }
        let broker = Self { db_path: path };
        broker.init_db()?;
        Ok(broker)
    }

    pub fn default_path() -> PathBuf {
        let home = std::env::var("HOME").unwrap_or_default();
        PathBuf::from(home).join(".cruxdev").join("bus.db")
    }

    fn conn(&self) -> anyhow::Result<Connection> {
        Ok(Connection::open(&self.db_path)?)
    }

    fn init_db(&self) -> anyhow::Result<()> {
        let conn = self.conn()?;
        conn.execute_batch(
            "CREATE TABLE IF NOT EXISTS sessions (
                id TEXT PRIMARY KEY,
                project TEXT NOT NULL,
                directory TEXT NOT NULL,
                registered_at REAL NOT NULL,
                last_heartbeat REAL NOT NULL
            );
            CREATE TABLE IF NOT EXISTS messages (
                id TEXT PRIMARY KEY,
                type TEXT NOT NULL,
                source_project TEXT NOT NULL,
                target_project TEXT NOT NULL,
                title TEXT NOT NULL,
                body TEXT NOT NULL,
                severity TEXT NOT NULL DEFAULT 'medium',
                created_at REAL NOT NULL,
                acknowledged INTEGER NOT NULL DEFAULT 0,
                acknowledged_at REAL
            );
            CREATE INDEX IF NOT EXISTS idx_messages_target
            ON messages(target_project, acknowledged);"
        )?;
        Ok(())
    }

    pub fn register_session(&self, project: &str, directory: &str) -> anyhow::Result<String> {
        let id = uuid::Uuid::new_v4().to_string()[..8].to_string();
        let now = now();
        self.conn()?.execute(
            "INSERT INTO sessions (id, project, directory, registered_at, last_heartbeat) VALUES (?1, ?2, ?3, ?4, ?5)",
            params![id, project, directory, now, now],
        )?;
        Ok(id)
    }

    pub fn list_sessions(&self, max_age_seconds: f64) -> anyhow::Result<Vec<Session>> {
        let cutoff = now() - max_age_seconds;
        let conn = self.conn()?;
        let mut stmt = conn.prepare(
            "SELECT id, project, directory, registered_at, last_heartbeat FROM sessions WHERE last_heartbeat > ?1 ORDER BY project"
        )?;
        let sessions = stmt.query_map(params![cutoff], |row| {
            Ok(Session {
                id: row.get(0)?,
                project: row.get(1)?,
                directory: row.get(2)?,
                registered_at: row.get(3)?,
                last_heartbeat: row.get(4)?,
            })
        })?.filter_map(|r| r.ok()).collect();
        Ok(sessions)
    }

    pub fn send_message(&self, msg_type: &str, source: &str, target: &str, title: &str, body: &str, severity: &str) -> anyhow::Result<String> {
        let id = uuid::Uuid::new_v4().to_string()[..8].to_string();
        self.conn()?.execute(
            "INSERT INTO messages (id, type, source_project, target_project, title, body, severity, created_at) VALUES (?1, ?2, ?3, ?4, ?5, ?6, ?7, ?8)",
            params![id, msg_type, source, target, title, body, severity, now()],
        )?;

        // Write notification file
        if target != "*" {
            self.write_notification(target, title);
        } else if let Ok(sessions) = self.list_sessions(3600.0) {
            for s in sessions {
                if s.project != source {
                    self.write_notification(&s.project, title);
                }
            }
        }
        Ok(id)
    }

    pub fn check_inbox(&self, project: &str) -> anyhow::Result<Vec<Message>> {
        let conn = self.conn()?;
        let mut stmt = conn.prepare(
            "SELECT id, type, source_project, target_project, title, body, severity, created_at, acknowledged FROM messages WHERE target_project IN (?1, '*') AND acknowledged = 0 ORDER BY created_at DESC"
        )?;
        let msgs = stmt.query_map(params![project], |row| {
            Ok(Message {
                id: row.get(0)?,
                msg_type: row.get(1)?,
                source_project: row.get(2)?,
                target_project: row.get(3)?,
                title: row.get(4)?,
                body: row.get(5)?,
                severity: row.get(6)?,
                created_at: row.get(7)?,
                acknowledged: row.get::<_, i32>(8)? != 0,
            })
        })?.filter_map(|r| r.ok()).collect();
        Ok(msgs)
    }

    pub fn acknowledge(&self, message_id: &str) -> anyhow::Result<bool> {
        let n = self.conn()?.execute(
            "UPDATE messages SET acknowledged = 1, acknowledged_at = ?1 WHERE id = ?2",
            params![now(), message_id],
        )?;
        Ok(n > 0)
    }

    // --- Notification files ---

    fn notify_dir(&self) -> PathBuf {
        self.db_path.parent().unwrap_or(Path::new(".")).join("notifications")
    }

    fn write_notification(&self, project: &str, title: &str) {
        let dir = self.notify_dir();
        let _ = fs::create_dir_all(&dir);
        let path = dir.join(format!("{project}.notify"));

        let count = fs::read_to_string(&path).ok()
            .and_then(|c| serde_json::from_str::<serde_json::Value>(&c).ok())
            .and_then(|v| v.get("count").and_then(|c| c.as_u64()))
            .unwrap_or(0);

        let data = serde_json::json!({"count": count + 1, "latest_title": title, "updated_at": now()});
        let tmp = format!("{}.tmp", path.display());
        if fs::write(&tmp, data.to_string()).is_ok() {
            let _ = fs::rename(&tmp, &path);
        }
    }

    pub fn read_notification(&self, project: &str) -> Option<serde_json::Value> {
        let path = self.notify_dir().join(format!("{project}.notify"));
        let content = fs::read_to_string(&path).ok()?;
        serde_json::from_str(&content).ok()
    }

    pub fn clear_notification(&self, project: &str) -> bool {
        let path = self.notify_dir().join(format!("{project}.notify"));
        fs::remove_file(&path).is_ok()
    }
}

fn now() -> f64 {
    std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .unwrap_or_default()
        .as_secs_f64()
}

#[cfg(test)]
mod tests {
    use super::*;

    fn test_broker() -> (tempfile::TempDir, Broker) {
        let dir = tempfile::tempdir().unwrap();
        let db = dir.path().join("bus.db");
        let broker = Broker::new(db.to_str().unwrap()).unwrap();
        (dir, broker)
    }

    #[test]
    fn register_and_list() {
        let (_dir, broker) = test_broker();
        broker.register_session("crux", "/crux").unwrap();
        broker.register_session("cruxdev", "/cruxdev").unwrap();
        let sessions = broker.list_sessions(3600.0).unwrap();
        assert_eq!(sessions.len(), 2);
    }

    #[test]
    fn send_and_receive() {
        let (_dir, broker) = test_broker();
        broker.send_message("issue", "crux", "cruxdev", "Bug", "details", "high").unwrap();
        let msgs = broker.check_inbox("cruxdev").unwrap();
        assert_eq!(msgs.len(), 1);
        assert_eq!(msgs[0].title, "Bug");
    }

    #[test]
    fn acknowledge_message() {
        let (_dir, broker) = test_broker();
        broker.send_message("issue", "crux", "cruxdev", "Bug", "d", "medium").unwrap();
        let msgs = broker.check_inbox("cruxdev").unwrap();
        assert!(broker.acknowledge(&msgs[0].id).unwrap());
        assert!(broker.check_inbox("cruxdev").unwrap().is_empty());
    }

    #[test]
    fn notification_files() {
        let (_dir, broker) = test_broker();
        broker.send_message("issue", "src", "target", "Bug", "d", "medium").unwrap();
        let notif = broker.read_notification("target");
        assert!(notif.is_some());
        assert_eq!(notif.unwrap()["count"], 1);
        assert!(broker.clear_notification("target"));
        assert!(broker.read_notification("target").is_none());
    }

    #[test]
    fn broadcast_notifies_others() {
        let (_dir, broker) = test_broker();
        broker.register_session("a", "/a").unwrap();
        broker.register_session("b", "/b").unwrap();
        broker.send_message("pattern", "a", "*", "Tip", "desc", "low").unwrap();
        assert!(broker.read_notification("b").is_some());
        assert!(broker.read_notification("a").is_none()); // sender excluded
    }
}
